### PR TITLE
kernel object validation infrastructure

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This implements kernel object validation as discussed in https://jira.zephyrproject.org/browse/ZEP-2187

The actual validation will take place in forthcoming patches to implement system calls, and only done upon privilege elevation.

Not all kernel object types are connected to this mechanism; forthcoming work will enable the rest  of them as well as device drivers.

### CODE REVIEW

File: kernel/object_validate.c [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * @param object Address of kernel object
+ * @param thread Thread to grant access to the object
+ */
+void k_object_grant_access(void *object, struct k_thread *thread)
+{
+	struct _k_object *ko = _k_object_find(object);
+
+	if (!ko) {
+		printk("granting access to non-existent kernel object %p\n",
```
Comment (lpereira) [inline kernel/object_validate.c]: Leftover debugging printk()?

Comment (andrewboie) [reply kernel/object_validate.c]: this is intentional, k_object_grant_access() is a public API and should informatively freak out if a bad object is passed in.

Comment (inakypg) [reply kernel/object_validate.c]: Something we need is a more "regular" way to print warning/error messages so they are easier to monitor for (not the subject of this patch though).

File: scripts/process_gperf.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            r"hash((const char *)&str, len)", line)
+
+    # Just compare pointers directly instead of using memcmp
+    if re.search("if [(][*]str", line):
+        fp.write("            if (str == s)\n")
+        return
+
+    # Take the strings with the binary information for the pointer values,
+    # and just turn them into pointers
+    line = re.sub(r'["].*["]', reformat_str, line)
```
Comment (lpereira) [inline scripts/process_gperf.py]: More things that could be done here:

* Remove the `len` parameter from `hash()` and `_k_object_lookup()`
* `len` can be added somewhere in `_k_object_lookup()` as a constant value of 4. This should get rid of some branches inside that function

File: scripts/gen_kobject_list.py [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+# a string, we told gperf to expect binary strings that are not NULL-terminated.
+footer = """%%
+struct _k_object *_k_object_find(void *obj)
+{
+    return _k_object_lookup((const char *)obj, sizeof(void *));
+}
+"""
+
+
+def write_gperf_table(fp, objs, static_begin, static_end):
```
Comment (lpereira) [inline scripts/gen_kobject_list.py]: Played around with this, and it seems we can reduce the table even further if we do some processing on these pointers. The caveat is that the lookup will be slightly slower; but worth checking out, maybe later:

- Subtract the smaller pointer value from the list of all the pointers.  In the obj_validation test for qemu_x86, this made the largest value keep well below what a 16-bit integer can hold
- Look at all the values if they're aligned: if the least significant nibble for all the pointers is `0`, these values can be shifted-right by 2. In the obj_validation test for qemu_x86, all the values were well below what an 8-bit integer can hold

The _k_object structure has to be changed based on these findings, and _k_object_find() has to undo the compression, but this reduces the overhead without too much work

File: kernel/object_validate.c (hunk 1 above)
Comment (lpereira) [reply kernel/object_validate.c]: This should be guarded by CONFIG_PRINTK anyway (other places of the code do this)

Comment (andrewboie) [reply kernel/object_validate.c]: that's not correct. printk() expands to a no-op inline function if CONFIG_PRINTK=n.
the reason why some other calls are guarded is because they reference the otype_to_str() function. this function is itself guarded because GCC's -ffunction/data-sections doesn't work the way you would expect with strings, and the strings would end up in the binary anyway even if otype_to_str was never used.

File: kernel/object_validate.c [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+				   struct k_thread *thread)
+{
+	/* STUB, implement */
+	ARG_UNUSED(ko);
+	ARG_UNUSED(thread);
+}
+
+
+/**
+ * grant a thread access to a kernek object
```
Comment (lpereira) [inline kernel/object_validate.c]: s/kernek/kernel/

File: scripts/process_gperf.py (hunk 2 above)
Comment (andrewboie) [reply scripts/process_gperf.py]: looking at the generated assembly code, the compiler appears to be smart enough to disregard len entirely since both hash and k_object_lookup are now static functions, called only from _k_object_find() which uses a constant parameter (4) for len,

I don't currently see a need for these two steps unless I'm missing something.

For example on x86, all of the code currently reduces to:

Dump of assembler code for function _k_object_find:
   0x00002b84 <+0>:	push   %ebp
   0x00002b85 <+1>:	mov    %esp,%ebp
   0x00002b87 <+3>:	push   %eax
   0x00002b88 <+4>:	mov    0x8(%ebp),%ecx
   0x00002b8b <+7>:	movzbl %cl,%eax
   0x00002b8e <+10>:	movzbl 0x38c0(%eax),%edx
   0x00002b95 <+17>:	xor    %eax,%eax
   0x00002b97 <+19>:	cmp    $0x1,%edx
   0x00002b9a <+22>:	ja     0x2bac <_k_object_find+40>
   0x00002b9c <+24>:	cmp    0x40d000(,%edx,8),%ecx
   0x00002ba3 <+31>:	jne    0x2bac <_k_object_find+40>
   0x00002ba5 <+33>:	lea    0x40d000(,%edx,8),%eax
   0x00002bac <+40>:	leave  
   0x00002bad <+41>:	ret    
End of assembler dump.

File: kernel/object_validate.c (hunk 4 above)
Comment (andrewboie) [reply kernel/object_validate.c]: fixed

**Review by dbkinder — DISMISSED** [review]
The only doc changes here is to add gperf to the development environment.  Will there be some documentation (eventually) that talks about the validation mechanism for user-supplied kernel object pointers?

**Comment by andrewboie** [comment]
I don't think it's useful to write documentation for features that are not user-facing. All of this is 'behind the scenes' stuff to support userspace.
There will be formal documentation on userspace/thread protection in the fullness of time, but we are a long way away from that.

**Comment by nashif** [comment]
> I don't think it's useful to write documentation for features that are not user-facing.

 I think a (slightly modified) subset of the proposal text you have in the Jira makes a good introduction of the object verification feature in an overall chapter we should have about memory protection, thread isolation and "user-space" features to be introduced later.

File: scripts/process_gperf.py (hunk 2 above)
Comment (lpereira) [reply scripts/process_gperf.py]: I've made that comment before looking at the assembly output. My bad. Yeah, those changes are unnecessary.

**Review by lpereira — DISMISSED** [review]

**Comment by lpereira** [comment]
About the documentation: yeah, I agree.  We say Zephyr is built with security in mind and having something like this in the documentation is helpful... specially since this isn't done in a common way.

**Comment by andrewboie** [comment]
Guys, it is *way* too early in the development of this feature to be writing documentation on how it works internally. Too many things can and will change as the design evolves.
Also I prefer such documentation to be extremely high level, if people want to know how something works in detail they need to look at the code, which should be well-commented.
I do not intend to start writing formal documentation for memory protection until implementation is almost complete.

**Comment by dbkinder** [comment]
I was hoping for an overview of requirements driving the implementation of various object verification capabilities (memory protection, thread isolation, etc.)  rather than the specific implementation details.

**Comment by andrewboie** [comment]
@dbkinder what you are looking for is documented in an RFC in the comments for the JIRA linked in the PR description.
Formal documentation under the doc/ directory will *not* be written until the overall mechanism is almost complete. If I write it now it will just have to be re-written later.

**Comment by nashif** [comment]
> I do not intend to start writing formal documentation for memory protection until implementation is almost complete.

That I fully agree with you on, I was not hoping for this documentation to be part of this PR or 1.9, should be part of the overall memory protection feature which is not complete yet for 1.9

File: kernel/object_validate.c [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	 */
+	if (!(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
+		_k_object_grant_access(ko, _current);
+	} else {
+		/* TODO: If this object was already initialized, and the caller
+		 * was from userspace, and doesn't have permissions, k_oops()
+		 */
+	}
+
+	memset(ko->perms, 0, CONFIG_MAX_THREAD_BITS);
```
Comment (chunlinhan) [inline kernel/object_validate.c]: In line 181, we just grant the access permission to the caller thread, but clear all permission bits here?

Comment (andrewboie) [reply kernel/object_validate.c]: yeah this isn't right. looks like i fixed it in some later work but I need to backport to this series.

**Comment by andrewboie** [comment]
New patch set differences:

- No changes to the python scripts or any of the stuff which generates and links the gperf tables
- CONFIG_KERNEL_OBJECT_VERIFY rolled into CONFIG_USERSPACE. Further work revealed that object validation doesn't make much sense on its own and it's tightly linked to userspace/system calls.
- kernel/object_validation.c renamed to userspace.c. Additional code to support userspace will go in there, not just object validation.
- Policy and logic fixes in kernel/userspace.c
- No longer turning this on by default for QEMU yet. Userspace implementation currently proceeding on x86, so we will just test that for now in tests/kernel/object_validation.

File: include/kernel.h [hunk 6]
```diff
@@ -123,6 +123,103 @@ struct k_timer;
 struct k_poll_event;
 struct k_poll_signal;
 
+enum k_objects {
+	K_OBJ_ALERT,
+	K_OBJ_DELAYED_WORK,
+	K_OBJ_MEM_SLAB,
```
Comment (andyross) [inline include/kernel.h]: Are slabs really needed?  Seems like (just like pools) the API should work just fine from user code when referring to memory inside userspace, and the utility of managing a kernel-side allocator from user code that can't touch it seems kinda weird.

File: kernel/userspace.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		return "k_mem_slab";
+	case K_OBJ_MSGQ:
+		return "k_msgq";
+	case K_OBJ_MUTEX:
+		return "k_mutex";
+	case K_OBJ_PIPE:
+		return "k_pipe";
+	case K_OBJ_SEM:
+		return "k_sem";
+	case K_OBJ_STACK:
```
Comment (andyross) [inline kernel/userspace.c]: Oh right, k_stack.  Remind me again why k_stack and k_fifo are different things (beyond the fact that one takes a void* and the other a u32_t)?  Not relevant to review, just something we should consider reviewing and unifying.

File: kernel/userspace.c [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * @param obj Address of kernel object to get metadata
+ * @return Kernel object's metadata, or NULL if the parameter wasn't the
+ * memory address of a kernel object
+ */
+__weak _GENERIC_SECTION(.kobject_data.text)
+struct _k_object *_k_object_find(void *obj)
+{
+	/* This is the __weak stub implementation, the version created by
+	 * gperf will replace it in build phase 2
```
Comment (andyross) [inline kernel/userspace.c]: FWIW: there's no need for this thing to actually be a function if it will never be called.  Linker names are typeless.  And no need to put it in a specific section if all you want is for the link to not fail.  Just a:

    __weak int _k_object_find;

...should be enough, no?  Or alternativesly a PROVIDES() directive in the linker script that is #if'd appropriately so it shows up only in the early linkage.

File: kernel/userspace.c [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	 */
+	ARG_UNUSED(obj);
+	return NULL;
+}
+
+#ifdef CONFIG_PRINTK
```
Comment (andyross) [inline kernel/userspace.c]: No need for this #if.  The function won't be linked if unreferenced, and it's not static so it won't generate an error.

File: kernel/userspace.c [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+			 */
+			return;
+		}
+	}
+
+	/* userspace can't grant access to objects unless it already has
+	 * access to that object
+	 */
+	if (_is_thread_user() && !test_thread_perms(ko)) {
+#ifdef CONFIG_PRINTK
```
Comment (andyross) [inline kernel/userspace.c]: Not sure if "PRINTK" is really the right predicate for "I want runtime error logging out of the memory protection layer".

Can't this be handled by the existing SYS_LOG_*() framework?

File: kernel/userspace.c [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+
+int _k_object_validate(void *obj, enum k_objects otype, int init)
+{
+	struct _k_object *ko;
+
+	ko = _k_object_find(obj);
+
+	if (!ko || ko->type != otype) {
+#ifdef CONFIG_PRINTK
```
Comment (andyross) [inline kernel/userspace.c]: Same complaint about printk vs. SYS_LOG

File: kernel/userspace.c [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	/* Uninitialized objects are not owned by anyone. However if an
+	 * object is initialized, and the caller is from userspace, then
+	 * we need to assert that the user thread has sufficient permissions
+	 * to re-initialize.
+	 */
+	if (ko->flags & K_OBJ_FLAG_INITIALIZED && _is_thread_user() &&
+	    !test_thread_perms(ko)) {
+#ifdef CONFIG_PRINTK
```
Comment (andyross) [inline kernel/userspace.c]: again

File: kernel/userspace.c [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+		       _current, otype_to_str(otype), obj);
+#endif
+		return -EPERM;
+	}
+
+	/* If we are not initializing an object, and the object is not
+	 * initialized, we should freak out
+	 */
+	if (!init && !(ko->flags & K_OBJ_FLAG_INITIALIZED)) {
+		printk("%p used before initialization\n", obj);
```
Comment (andyross) [inline kernel/userspace.c]: Oddly here's a situation where printk() is not wrapped in an #if.

File: kernel/userspace.c [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+	ko = _k_object_find(object);
+	if (!ko) {
+		/* Supervisor threads can ignore rules about kernel objects
+		 * and may declare them on stacks, etc. Such objects will never
+		 * be usable from userspace, but we shouldn't explode.
+		 */
+		return;
+	}
+
+	memset(ko->perms, 0, CONFIG_MAX_THREAD_BITS);
```
Comment (andyross) [inline kernel/userspace.c]: Naming mistake or size goof?  If that config value is in bits, then this is clobbering 8x too much memory.  If it's not, it should be renamed.

File: scripts/gen_kobject_list.py [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+
+        # Is this a kernel object, or a structure containing kernel objects?
+        if type_offset not in type_env:
+            continue
+
+        if "DW_AT_declaration" in die.attributes:
+            # FIXME: why does k_sys_work_q not resolve an address in the DWARF
+            # data??? Every single instance it finds is an extern definition
+            # but not the actual instance in system_work_q.c
+            # Is there something weird about how lib-y stuff is linked?
```
Comment (andyross) [inline scripts/gen_kobject_list.py]: Yes, actually: the kernel directory is always presented to the link as an already-linked static library, to prevent the case where pulling symbols from one file pulls in static data too and bloats the image.  It looks like k_sys_work_q is only ever referenced from inside the kernel, so that may be triggering some variant DWARF output.   How does this look in e.g. gdb?  Can you see the type there?

File: scripts/process_gperf.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+# --- debug stuff ---
+
+"""
+gperf C file post-processor
+
+We use gperf to build up a perfect hashtable of pointer values. The way gperf
+does this is to create a table 'wordlist' indexed by a string repreesentation
+of a pointer address, and then doing memcmp() on a string passed in for
+comparison
```
Comment (andyross) [inline scripts/process_gperf.py]: This is super clever, but I just have to whine:

How robust is this vs. changes to the code output in upstream gperf?  I doubt they're doing much development these days, but if they do it's not unlikely that this script will break badly.  It might be a good idea to put a version check somewhere (maybe right here: is there a "generated by gperv v${WHATEVER}" comment in the file?) to catch that circumstance and force us to revalidate the metacompiler.

File: kernel/userspace.c [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+static int _is_thread_user(void)
+{
+	/* STUB */
+
+	return 0;
+}
+
+void k_object_grant_access(void *object, struct k_thread *thread)
```
Comment (andyross) [inline kernel/userspace.c]: Design note: would be really nice to have a way to do this statically, given that the objects and threads are static objects.  Would be even nicer to have it be implicit, along the lines of "a static thread has access to any static kernel object declared in the same C source file"

File: scripts/sanitycheck [hunk 18]
```diff
@@ -476,13 +476,14 @@ class SizeCalculator:
...
-		   "_k_mem_slab_area", "_k_mem_pool_area",
-		   "_k_sem_area", "_k_mutex_area", "_k_alert_area",
-		   "_k_fifo_area", "_k_lifo_area", "_k_stack_area",
-		   "_k_msgq_area", "_k_mbox_area", "_k_pipe_area",
+                   "_static_thread_area", "_k_timer_area", "_k_work_area",
+                   "_k_mem_slab_area", "_k_mem_pool_area",
+                   "_k_sem_area", "_k_mutex_area", "_k_alert_area",
+                   "_k_fifo_area", "_k_lifo_area", "_k_stack_area",
+                   "_k_msgq_area", "_k_mbox_area", "_k_pipe_area",
                    "net_if", "net_if_event", "net_stack", "net_l2_data",
```
Comment (andyross) [inline scripts/sanitycheck]: Unrelated whitespace muckery snuck into the patch.

**Review by andyross — DISMISSED** [review]
A few notes & whines.  Seems surprisingly small given what it does!

File: include/kernel.h (hunk 6 above)
Comment (andrewboie) [reply include/kernel.h]: in its current implementation, the mem_slab APIs will not work from userspace. they contain wait queues, they have other threads pending on them, etc...redesign would be needed to make them work entirely from userspace. have a look at the code for k_mem_slab_free() there's all kinds of stuff in there that userspace can't do.

also its worth noting that detailed analysis of kernel object APIs is for a later stage in development of this feature, currently I am not changing any of them and just trying to get the infrastructure in place.

File: kernel/userspace.c (hunk 9 above)
Comment (andrewboie) [reply kernel/userspace.c]: yes, there is.
gc-sections is broken with C strings. even if gc-sections strips this function, the strings will still be in the binary :(
upstream fixed it, but i believe the fix landed in the GCC 7.x branch.

File: kernel/userspace.c (hunk 7 above)
Comment (andrewboie) [reply kernel/userspace.c]: yeah we have some issues with these APIs, additionally see https://jira.zephyrproject.org/browse/ZEP-2237

File: kernel/userspace.c (hunk 17 above)
Comment (andrewboie) [reply kernel/userspace.c]: i'd like to punt on this for now, but that could be a future enhancement once all the implications have been analyzed.

File: kernel/userspace.c (hunk 10 above)
Comment (andrewboie) [reply kernel/userspace.c]: these ifdefs are only here to prevent build errors from use of otype_to_str() when CONFIG_PRINTK is turned off (and hence the function doesn't exist)

File: scripts/gen_kobject_list.py (hunk 15 above)
Comment (andrewboie) [reply scripts/gen_kobject_list.py]: I can't seem to find it in the DWARF output at all except as an extern reference. It's really baffling. I'm going to file a bug for further investigation.

File: scripts/process_gperf.py (hunk 16 above)
Comment (andrewboie) [reply scripts/process_gperf.py]: good idea, I'll add a version string check and error out. so far this works in gperf 3.0.x and 3.1. agree this will fall over if the generated C code changes too much.

File: kernel/userspace.c (hunk 17 above)
Comment (andrewboie) [reply kernel/userspace.c]: https://jira.zephyrproject.org/browse/ZEP-2585 to look into this

File: kernel/userspace.c (hunk 8 above)
Comment (andrewboie) [reply kernel/userspace.c]: i'll give that a try!

File: tests/kernel/gen_isr_table/prj.conf [hunk 19]
```diff
@@ -4,4 +4,4 @@
 # A memory barrier does not help, we need an 'instruction barrier' but
 # GCC doesn't support this; we need to tell the compiler not to reorder
 # memory accesses to trigger_check around calls to trigger_irq.
-CONFIG_COMPILER_OPT="-O0"
+CONFIG_DEBUG=y
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply tests/kernel/gen_isr_table/prj.conf]: attribute((optimize("O0"))) on the relevant function sounds better, I'll do that.

File: kernel/userspace.c (hunk 14 above)
Comment (andrewboie) [reply kernel/userspace.c]: oops, naming sucks, will fix this

File: tests/kernel/gen_isr_table/prj.conf (hunk 19 above)
[the remark this answers is not in the document]
Comment (lpereira) [reply tests/kernel/gen_isr_table/prj.conf]: Even then, will `O0` guarantee there won't be any instruction reordering?

**Comment by andrewboie** [comment]
updated patch set:

- sanitycheck whitespace changes moved into its own patch
- gperf version checking in post-processing script
- __attribute__((optimize("-O0"))) used in gen_isr_table test, plus some other fixes
- use linker PROVIDE() for _k_object_find() instead of dummy function
- CONFIG_MAX_THREAD_BITS now named CONFIG_MAX_THREAD_BYTES
- make use of CONFIG_PRINTK in userspace.c much clearer

File: tests/kernel/gen_isr_table/src/main.c [hunk 20]
```diff
@@ -98,6 +98,15 @@ void isr4(void *param)
...
 
+/* Need to turn optimization off. Otherwise compiler may generate incorrect
+ * code, not knowing that trigger_irq() affects the value of trigger_check,
+ * even if declared volatile.
+ *
+ * A memory barrier does not help, we need an 'instruction barrier' but GCC
+ * doesn't support this; we need to tell the compiler not to reorder memory
+ * accesses to trigger_check around calls to trigger_irq.
+ */
+__attribute__((optimize("-O0")))
```
Comment (agross-linaro) [inline tests/kernel/gen_isr_table/src/main.c]: This idea of using the optimize to somehow give an instruction barrier makes me uneasy.  Why not just add a __ISB() for ARM architectures.   I am sure other arches have similar solutions.

File: scripts/sanitycheck (hunk 18 above)
Comment (agross-linaro) [reply scripts/sanitycheck]: doesn't the checkpatch script catch that?

**Review by agross-oss — COMMENTED** [review]
The only questionable thing I saw was the optimization usage.  I'd be more comfortable with a std instruction/data barrier being used.

File: scripts/sanitycheck (hunk 18 above)
Comment (andrewboie) [reply scripts/sanitycheck]: @agross-linaro catch what, exactly?
at any rate, the whitespace changes are in their own patch in the series.

File: tests/kernel/gen_isr_table/src/main.c (hunk 20 above)
Comment (andrewboie) [reply tests/kernel/gen_isr_table/src/main.c]: i don't want to get too sidetracked by this -- I needed to get rid of the global -O0 here because it messes up the padding size for the generated gperf code in the linker script.

I moved it to the function instead. if we want to further iterate on it, please file a bug.

**Comment by andrewboie** [comment]
@agross-linaro the optimize thing was already there, its really not in the scope of this patch, i'm just moving it. can you file a bug if you want it looked at further?

**Comment by agross-linaro** [comment]
Fair enough.  I'll write one up.

**Review by andyross — APPROVED** [review]
Can't see anything else worth whining about.

Note that (now that I'm thinkinga bout the root issue and not just how to do optimization pragmas) it's probably possible to get gcc to do the right thing with a noop in a asm() statement that declares memory as a clobber.  But agreed, that's a separate problem.
